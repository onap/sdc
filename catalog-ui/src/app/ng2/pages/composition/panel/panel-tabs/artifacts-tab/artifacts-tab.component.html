<div class="w-sdc-designer-sidebar-tab-content artifacts">
    <div class="w-sdc-designer-sidebar-section">
        <ng2-expand-collapse state="0">
            <header sdc-tooltip tooltip-text="{{title}}">{{title}}</header>
            <content class="artifacts-container">
                <div class="w-sdc-designer-sidebar-section-content">
                    <div class="i-sdc-designer-sidebar-section-content-item" *ngFor="let artifact of artifacts$ | async">
                        <div class="i-sdc-designer-sidebar-section-content-item-artifact"
                             *ngIf="(!isComponentInstanceSelected || artifact.esId) && 'HEAT_ENV' !== artifact.artifactType"
                             attr.data-tests-id="'artifact-item-' + artifact.artifactDisplayName">
                            <span *ngIf="artifact.heatParameters?.length"
                                  class="i-sdc-designer-sidebar-section-content-item-file-link"></span>
                            <div class="i-sdc-designer-sidebar-section-content-item-artifact-details"
                                 [class.heat]="artifact.isHEAT() && artifact.heatParameters?.length">
                                <div *ngIf="artifact.artifactName"
                                     class="i-sdc-designer-sidebar-section-content-item-artifact-filename"
                                     attr.data-tests-id="artifactName-{{artifact.artifactDisplayName}}"
                                     sdc-tooltip tooltip-text="{{artifact.artifactName}}">{{artifact.artifactName}}
                                </div>
                                <div class="artifact-buttons-container upper-buttons">


                                    <svg-icon
                                            *ngIf="!isViewOnly && !artifact.isFromCsar && artifact.artifactName"
                                            name="trash-o" clickable="true" size="medium" mode="info"
                                            class="artifact-button" testId="delete_{{artifact.artifactDisplayName}}"
                                            (click)="delete(artifact)"></svg-icon>

                                    <!--Display env parameters edit button for Instance -->
                                    <svg-icon
                                            *ngIf="!isViewOnly  && artifact.isHEAT() && isComponentInstanceSelected && artifact.heatParameters?.length"
                                            name="indesign_status" clickable="true" size="medium" mode="info"
                                            class="artifact-button"
                                            testId="edit-parameters-of-{{artifact.artifactDisplayName}}"
                                            (click)="updateEnvParams(artifact)"
                                            tooltip="Edit ENV Params"
                                    ></svg-icon>

                                    <!--Display env parameters VIEW button for Instance -->
                                    <svg-icon
                                            *ngIf="isViewOnly && artifact.isHEAT() && isComponentInstanceSelected && artifact.heatParameters?.length"
                                            name="inputs-o" clickable="true" size="medium" mode="info"
                                            class="artifact-button"
                                            testId="view-parameters-of-{{artifact.artifactDisplayName}}"
                                            (click)="viewEnvParams(artifact)"
                                            tooltip="View ENV Params"
                                    ></svg-icon>

                                    <!--Display env parameters edit button for VF -->
                                    <svg-icon
                                            *ngIf = "!isViewOnly && !isComponentInstanceSelected && artifact.heatParameters?.length"
                                            name="indesign_status" clickable="true" size="medium" mode="info"
                                            class="artifact-button"
                                            testId="edit-parameters-of-{{artifact.artifactDisplayName}}"
                                            (click)="updateEnvParams(artifact)"></svg-icon>


                                    <download-artifact *ngIf="artifact.esId && 'deployment' != type"
                                                       class="artifact-button"
                                                       [artifact]="artifact" [componentType]="component.componentType"
                                                       [componentId]="component.uniqueId"
                                                       testId="download_{{artifact.artifactDisplayName}}"
                                                       [isInstance]="isComponentInstanceSelected"></download-artifact>
                                    <download-artifact *ngIf="artifact.esId && 'deployment' == type"
                                                       class="artifact-button"
                                                       [artifact]="artifact" [componentType]="component.componentType"
                                                       [componentId]="component.uniqueId"
                                                       [isInstance]="isComponentInstanceSelected"
                                                       testId="download_{{artifact.artifactDisplayName}}"
                                                       [showLoader]="artifact.isHEAT()"></download-artifact>

                                    <button *ngIf="!isViewOnly && !artifact.esId && type==='deployment' && !isComponentInstanceSelected && !artifact.isThirdParty()"
                                            class="artifact-button attach sprite e-sdc-small-icon-upload"
                                            (click)="addOrUpdate(artifact)" type="button"
                                            attr.data-tests-id="add_Artifact"></button>
                                </div>
                                <div>
                                <span class="i-sdc-designer-sidebar-section-content-item-artifact-details-name"
                                      attr.data-tests-id="artifact_Display_Name-{{artifact.artifactDisplayName}}"
                                      [ngClass]="{'hand enabled': artifact.allowDeleteAndUpdate}"
                                      (click)="artifact.allowDeleteAndUpdate && addOrUpdate(artifact)"
                                      sdc-tooltip tooltip-text="{{artifact.artifactDisplayName}}">{{artifact.artifactDisplayName}}</span>
                                    <div class="i-sdc-designer-sidebar-section-content-item-artifact-heat-env"
                                         *ngIf="artifact.heatParameters?.length">
                                        <span attr.data-tests-id="heat_env_{{artifact.artifactDisplayName}}">{{artifact.artifactDisplayName}} (ENV)</span>
                                        <div class="artifact-buttons-container">
                                            <svg-icon *ngIf="!isViewOnly && envArtifactOf(artifact)"
                                                      name="edit-o" clickable="true" size="medium"
                                                      mode="info" class="artifact-button edit-pencil"
                                                      testId="edit_{{artifact.artifactDisplayName}}"
                                                      (click)="addOrUpdate(envArtifactOf(artifact))"></svg-icon>

                                            <download-artifact [artifact]="envArtifactOf(artifact)"
                                                               class="artifact-button"
                                                               [componentType]="component.componentType"
                                                               [componentId]="component.uniqueId"
                                                               [isInstance]="isComponentInstanceSelected"
                                                               testId="download_env_{{artifact.artifactDisplayName}}"></download-artifact>
                                        </div>
                                    </div>
                                </div>

                                <div class="i-sdc-designer-sidebar-section-content-item-artifact-details-desc">
                                    <span class="i-sdc-designer-sidebar-section-content-item-artifact-details-desc-label"
                                          *ngIf="artifact.description">Description:</span>{{artifact.description}}
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                <div class="w-sdc-designer-sidebar-section-footer"
                     *ngIf="!isViewOnly && type!=='api' && (!isComponentInstanceSelected || isVfOrPnf() && (type !== 'deployment') || isComplex)">
                    <sdc-button testId="add_Artifact_Button" size="large" type="primary" text="Add Artifact"
                                (click)="addOrUpdate({})"></sdc-button>
                </div>
            </content>
        </ng2-expand-collapse>
    </div>
</div>
