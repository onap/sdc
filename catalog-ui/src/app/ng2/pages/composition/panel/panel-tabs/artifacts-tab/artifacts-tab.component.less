@import '../../../../../../../assets/styles/override';


.artifacts /deep/ .expand-collapse-content {
    padding: 10px 0px;

    &.collapsed {
        padding: 0 0;
    }
}

.i-sdc-designer-sidebar-section-content-item-artifact {

    &:not(:hover) .artifact-button {
        display:none;
    }
    .artifact-buttons-container {
        display: inline-flex;
        flex-direction: row-reverse;
        position: absolute;
        right:0;

        &.upper-buttons {
            margin-top: 8px;
        }

        .artifact-button {
            cursor:pointer;
            padding-right:5px;

            &.edit-pencil {
                margin-top: 10px;
            }
        }
    }
}

.w-sdc-designer-sidebar-section-footer {
    padding: 20px;
    display: flex;
    justify-content: center;
    
}


.w-sdc-designer-sidebar-tab-content.artifacts {

    .i-sdc-designer-sidebar-section-content-item-artifact.hand {
      cursor: pointer;
    }

    .w-sdc-designer-sidebar-section-content {
        padding: 0;
    }
    .w-sdc-designer-sidebar-section-title {
        &.expanded {
            margin-bottom: 0;
        }
    }

    .i-sdc-designer-sidebar-section-content-item-artifact-details {
        display: inline-block;
        margin-left: 5px;
        vertical-align: middle;
        width: 180px;
        &.heat {
            line-height: 18px;
            width: 250px;
        }
    }

    .i-sdc-designer-sidebar-section-content-item-artifact-details-name {

        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
        max-width:220px;
        display: inline-block;
        //text-transform: capitalize;
        &.enabled {
            &:hover {
                color: @sdcui_color_dark-blue;
            }
        }

    }

    .i-sdc-designer-sidebar-section-content-item-artifact-heat-env {
        color: @sdcui_color_dark-gray;
        margin-top: 6px;
        line-height: 42px;
        padding-top: 10px;
        border-top:1px solid #c8cdd1;
        .enabled {
            &:hover {
                cursor: pointer;
                color: @sdcui_color_dark-blue;
            }
        }
    }

    .i-sdc-designer-sidebar-section-content-item-artifact-filename {
        color: @sdcui_color_dark-gray;
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
        max-width: 225px;
        display: inline-block;
        font-weight: bold;
        &.enabled {
            &:hover {
                color: @sdcui_color_dark-blue;
            }
        }
    }


    .i-sdc-designer-sidebar-section-content-item-file-link{
        border-left: 1px #848586 solid;
        height: 58px;
        margin-left: -11px;
        margin-top: 11px;
        border-top: 1px #848586 solid;
        border-bottom: 1px #848586 solid;
        width: 12px;
        float: left;
    }

    .i-sdc-designer-sidebar-section-content-item-artifact-details-desc {
        display: none;
        line-height: 16px;
        word-wrap: break-word;
        white-space: normal;
    }

    .i-sdc-designer-sidebar-section-content-item-artifact-details-desc-label {
        color: @sdcui_color_dark-gray;
    }


    .i-sdc-designer-sidebar-section-content-item-artifact {
        border-bottom: 1px solid #c8cdd1;
        padding: 5px 10px 5px 18px;
        position: relative;
        // line-height: 36px;
        min-height: 61px;
        //cursor: default;
        display: flex;
        align-items: center;


        .i-sdc-designer-sidebar-section-content-item-button {
            top: 20px;
            line-height: 10px;
        }

        &:hover {
            //background-color: @color_c;
            background-color: white;
            transition: all .3s;

            .i-sdc-designer-sidebar-section-content-item-button {
                display: block;

            }

        }
    }
}
