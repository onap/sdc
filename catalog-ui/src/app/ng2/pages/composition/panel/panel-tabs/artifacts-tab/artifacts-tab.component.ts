import { Component, Input } from '@angular/core';
import { Store } from '@ngxs/store';
import { ArtifactModel, Component as TopologyTemplate, FullComponentInstance, Resource } from 'app/models';
import { WorkspaceService } from 'app/ng2/pages/workspace/workspace.service';
import { ResourceNamePipe } from 'app/ng2/pipes/resource-name.pipe';
import { ComponentInstanceServiceNg2 } from 'app/ng2/services/component-instance-services/component-instance.service';
import { TopologyTemplateService } from 'app/ng2/services/component-services/topology-template.service';
import { ArtifactType } from 'app/utils';
import * as _ from 'lodash';
import { SdcUiServices } from 'onap-ui-angular';
import { Observable } from 'rxjs/Observable';
import { map } from 'rxjs/operators';
import { ArtifactsService } from '../../../../../components/forms/artifacts-form/artifacts.service';
import { GetArtifactsByTypeAction } from '../../../../../store/actions/artifacts.action';
import { GetInstanceArtifactsByTypeAction } from '../../../../../store/actions/instance-artifacts.actions';
import { ArtifactsState } from '../../../../../store/states/artifacts.state';
import { InstanceArtifactsState } from '../../../../../store/states/instance-artifacts.state';
import { SelectedComponentType, TogglePanelLoadingAction } from '../../../common/store/graph.actions';
import { CompositionService } from '../../../composition.service';

@Component({
    selector: 'artifacts-tab',
    styleUrls: ['./artifacts-tab.component.less'],
    templateUrl: './artifacts-tab.component.html',
    providers: [SdcUiServices.ModalService]
})

export class ArtifactsTabComponent {

    @Input() component: FullComponentInstance | TopologyTemplate;
    @Input() isViewOnly: boolean;
    @Input() input: any;
    @Input() componentType: SelectedComponentType;

    public title: string;
    public type: string;
    public isComponentInstanceSelected: boolean;
    public artifacts$: Observable<ArtifactModel[]>;
    private topologyTemplateType: string;
    private topologyTemplateId: string;
    private heatToEnv: Map<string, ArtifactModel>;
    private resourceType: string;
    private isComplex: boolean;

    constructor(private store: Store,
                private compositionService: CompositionService,
                private workspaceService: WorkspaceService,
                private componentInstanceService: ComponentInstanceServiceNg2,
                private topologyTemplateService: TopologyTemplateService,
                private artifactService: ArtifactsService) {
        this.heatToEnv = new Map();
    }

    ngOnInit() {
        this.topologyTemplateType = this.workspaceService.metadata.componentType;
        this.topologyTemplateId = this.workspaceService.metadata.uniqueId;
        this.type = this.input.type;
        this.title = this.getTitle(this.type);
        this.isComponentInstanceSelected = this.componentType === SelectedComponentType.COMPONENT_INSTANCE;
        this.resourceType = this.component['resourceType'];
        this.isComplex = this.component.isComplex();
        this.loadArtifacts();
    }

    public addOrUpdate = (artifact: ArtifactModel): void => {
        if (this.isComponentInstanceSelected) {
            this.artifactService.openArtifactModal(this.topologyTemplateId, this.topologyTemplateType, artifact, this.type, this.isViewOnly, this.component.uniqueId);
        } else {
            this.artifactService.openArtifactModal(this.topologyTemplateId, this.topologyTemplateType, artifact, this.type, this.isViewOnly);
        }
    }

    public updateEnvParams = (artifact: ArtifactModel) => {
        if (this.isComponentInstanceSelected) {
            this.artifactService.openUpdateEnvParams(this.topologyTemplateType, this.topologyTemplateId, this.heatToEnv.get(artifact.uniqueId), this.component.uniqueId);
        } else {
            this.artifactService.openUpdateEnvParams(this.topologyTemplateType, this.topologyTemplateId, artifact);
        }
    }

    public viewEnvParams = (artifact: ArtifactModel) => {
        if (this.isComponentInstanceSelected) {
            this.artifactService.openViewEnvParams(this.topologyTemplateType, this.topologyTemplateId, this.heatToEnv.get(artifact.uniqueId), this.component.uniqueId);
        } else {
            this.artifactService.openViewEnvParams(this.topologyTemplateType, this.topologyTemplateId, artifact);
        }
    }

    public getEnvArtifact = (heatArtifact: ArtifactModel, artifacts: ArtifactModel[]): ArtifactModel => {
        const envArtifact = _.find(artifacts, (item: ArtifactModel) => {
            return item.generatedFromId === heatArtifact.uniqueId;
        });
        if (envArtifact && heatArtifact) {
            envArtifact.artifactDisplayName = heatArtifact.artifactDisplayName;
            envArtifact.timeout = heatArtifact.timeout;
        }
        return envArtifact;
    }

    public delete = (artifact: ArtifactModel): void => {
        if (this.isComponentInstanceSelected) {
            this.artifactService.deleteArtifact(this.topologyTemplateType, this.topologyTemplateId, artifact, this.component.uniqueId);
        } else {
            this.artifactService.deleteArtifact(this.topologyTemplateType, this.topologyTemplateId, artifact);
        }
    }

    public isVfOrPnf(): boolean {
        if (this.component.isResource()){
            if (this.resourceType) {
                return this.resourceType === 'VF' || this.resourceType == 'PNF';
            }
            return false;
        }

        return false;
    }

    private envArtifactOf(artifact: ArtifactModel): ArtifactModel {
        return this.heatToEnv.get(artifact.uniqueId);
    }

    private isLicenseArtifact = (artifact: ArtifactModel): boolean => {
        let isLicense: boolean = false;
        if (this.component.isResource && (this.component as Resource).isCsarComponent) {
            if (ArtifactType.VENDOR_LICENSE === artifact.artifactType || ArtifactType.VF_LICENSE === artifact.artifactType) {
                isLicense = true;
            }
        }

        return isLicense;
    }

    private getTitle = (artifactType: string): string => {
        switch (artifactType) {
            case ArtifactType.SERVICE_API:
                return 'API Artifacts';
            case ArtifactType.DEPLOYMENT:
                return 'Deployment Artifacts';
            case ArtifactType.INFORMATION:
                return 'Informational Artifacts';
            default:
                return ResourceNamePipe.getDisplayName(artifactType) + ' Artifacts';
        }
    }

    private loadArtifacts = (forceLoad?: boolean): void => {

        this.store.dispatch(new TogglePanelLoadingAction({isLoading: true}));

        let action;
        if (this.isComponentInstanceSelected) {
            action = new GetInstanceArtifactsByTypeAction(({
                componentType: this.topologyTemplateType,
                componentId: this.topologyTemplateId,
                instanceId: this.component.uniqueId,
                artifactType: this.type
            }));
        } else {
            action = new GetArtifactsByTypeAction({
                componentType: this.topologyTemplateType,
                componentId: this.topologyTemplateId,
                artifactType: this.type
            });
        }
        this.store.dispatch(action).subscribe(() => {
            const stateSelector = this.isComponentInstanceSelected ? InstanceArtifactsState.getArtifactsByType : ArtifactsState.getArtifactsByType;
            this.artifacts$ = this.store.select(stateSelector).pipe(map((filterFn) => filterFn(this.type))).pipe(map((artifacts) => {
                _.forEach(artifacts, (artifact: ArtifactModel): void => {
                    const envArtifact = this.getEnvArtifact(artifact, artifacts); // Extract the env artifact (if exist) of the HEAT artifact
                    if (envArtifact) {
                        // Set a mapping between HEAT to HEAT_ENV
                        this.heatToEnv.set(artifact.uniqueId, envArtifact);
                    }
                });
                return _.orderBy(artifacts, ['mandatory', 'artifactDisplayName'], ['desc', 'asc']);
            }));

            this.artifacts$.subscribe((artifacts) => {
                _.forEach(artifacts, (artifact: ArtifactModel) => {
                    artifact.allowDeleteAndUpdate = this.allowDeleteAndUpdateArtifact(artifact);
                });
                if (this.component instanceof FullComponentInstance) {
                    this.compositionService.updateInstance(this.component);
                }
            });
            this.store.dispatch(new TogglePanelLoadingAction({isLoading: false}));
        }, () => {
            this.store.dispatch(new TogglePanelLoadingAction({isLoading: false}));
        });
    }

    private allowDeleteAndUpdateArtifact = (artifact: ArtifactModel): boolean => {
        if (!this.isViewOnly) {
            if (artifact.artifactGroupType === ArtifactType.DEPLOYMENT) {
                return !artifact.isFromCsar;
            } else {

                return (!artifact.isHEAT() && !artifact.isThirdParty() && !this.isLicenseArtifact(artifact));
            }
        }
        return false;
    }
}
