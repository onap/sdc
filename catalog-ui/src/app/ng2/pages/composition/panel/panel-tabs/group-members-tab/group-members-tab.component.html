<!--
  ~ Copyright (C) 2018 AT&T Intellectual Property. All rights reserved.
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~      http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->
 
<h1 class="w-sdc-designer-sidebar-section-title" tooltip="Members">Members
    <svg-icon-label *ngIf="!isViewOnly"
                    class="add-members-btn"
                    name="plus-circle-o"
                    mode="primary"
                    size="medium"
                    label="ADD"
                    labelPlacement="right"
                    (click)="openAddMembersModal()">
    </svg-icon-label>
</h1>
<div class="expand-collapse-content">
    <ul>
        <li *ngFor="let member of members; let i = index" class="component-details-panel-large-item"
            tooltip="{{member.name}}">
            <span>{{member.name}}</span>
            <svg-icon-label *ngIf="!isViewOnly"
                            name="trash-o" 
                            clickable="true" 
                            size="small" 
                            class="component-details-panel-item-delete"
                            data-tests-id="delete_member" 
                            (click)="deleteMember(member)"></svg-icon-label>
        </li>
    </ul>

    <div *ngIf="!members || members.length===0" class="component-details-panel-tab-no-data">
        <div class="component-details-panel-tab-no-data-title">No data to display yet</div>
        <div class="component-details-panel-tab-no-data-content">Add members to group to see members</div>
    </div>
</div>
