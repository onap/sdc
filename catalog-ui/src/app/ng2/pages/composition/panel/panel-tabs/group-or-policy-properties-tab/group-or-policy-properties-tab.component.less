@import '../../../../../../../assets/styles/override';

.group-policy-properties{
    display: block;
}