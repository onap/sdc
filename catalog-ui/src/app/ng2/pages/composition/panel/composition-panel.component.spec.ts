import { NO_ERRORS_SCHEMA } from '@angular/core';
import { async, ComponentFixture } from '@angular/core/testing';
import { NgxsModule, Store } from '@ngxs/store';
import { Observable } from 'rxjs';
import { Mock } from 'ts-mockery';
import { ConfigureFn, configureTests } from '../../../../../jest/test-config.helper';
import { Service } from '../../../../models/components/service';
import { Resource } from '../../../../models/components/resource';
import { GroupInstance } from '../../../../models/graph/zones/group-instance';
import { PolicyInstance } from '../../../../models/graph/zones/policy-instance';
import { ArtifactGroupType, ResourceType } from '../../../../utils/constants';
import { WorkspaceState } from '../../../store/states/workspace.state';
import { CompositionPanelComponent } from './composition-panel.component';
import { ArtifactsTabComponent } from './panel-tabs/artifacts-tab/artifacts-tab.component';
import { GroupMembersTabComponent } from './panel-tabs/group-members-tab/group-members-tab.component';
import { GroupOrPolicyPropertiesTab } from './panel-tabs/group-or-policy-properties-tab/group-or-policy-properties-tab.component';
import { InfoTabComponent } from './panel-tabs/info-tab/info-tab.component';
import { PolicyTargetsTabComponent } from './panel-tabs/policy-targets-tab/policy-targets-tab.component';
import { PropertiesTabComponent } from './panel-tabs/properties-tab/properties-tab.component';
import { ReqAndCapabilitiesTabComponent } from './panel-tabs/req-capabilities-tab/req-capabilities-tab.component';

describe('composition-panel component', () => {

    let fixture: ComponentFixture<CompositionPanelComponent>;
    let store: Store;

    const tabs = {
            infoTab : {titleIcon: 'info-circle', component: InfoTabComponent, input: {}, isActive: true, tooltipText: 'Information'},
            policyProperties: {
                titleIcon: 'settings-o', component: GroupOrPolicyPropertiesTab, input: {type: 'policy'}, isActive: false, tooltipText: 'Properties'
            },
            policyTargets: {titleIcon: 'inputs-o', component: PolicyTargetsTabComponent, input: {}, isActive: false, tooltipText: 'Targets'},
            groupMembers: {titleIcon: 'inputs-o', component: GroupMembersTabComponent, input: {}, isActive: false, tooltipText: 'Members'},
            groupProperties: {
                titleIcon: 'settings-o', component: GroupOrPolicyPropertiesTab, input: {type: 'group'}, isActive: false, tooltipText: 'Properties'
            },
            deploymentArtifacts: {
                titleIcon: 'deployment-artifacts-o', component: ArtifactsTabComponent,
                input: { type: ArtifactGroupType.DEPLOYMENT}, isActive: false, tooltipText: 'Deployment Artifacts'
            },
            apiArtifacts: {
                titleIcon: 'api-o', component: ArtifactsTabComponent,
                input: { type:  ArtifactGroupType.SERVICE_API}, isActive: false, tooltipText: 'API Artifacts'
            },
            infoArtifacts: {
                titleIcon: 'info-square-o', component: ArtifactsTabComponent,
                input: { type: ArtifactGroupType.INFORMATION}, isActive: false, tooltipText: 'Information Artifacts'
            },
            properties: {
                titleIcon: 'settings-o', component: PropertiesTabComponent,
                input: {title: 'Properties and Attributes'}, isActive: false, tooltipText: 'Properties'
            },
            reqAndCapabilities : {
                titleIcon: 'req-capabilities-o', component: ReqAndCapabilitiesTabComponent, input: {},
                isActive: false, tooltipText: 'Requirements and Capabilities'
            },
            inputs: {titleIcon: 'inputs-o', component: PropertiesTabComponent, input: {title: 'Inputs'}, isActive: false, tooltipText: 'Inputs'},
            settings: {titleIcon: 'settings-o', component: PropertiesTabComponent, input: {}, isActive: false, tooltipText: 'Settings'},
        };

    beforeEach(
        async(() => {

            const configure: ConfigureFn = (testBed) => {
                testBed.configureTestingModule({
                    declarations: [CompositionPanelComponent],
                    imports: [NgxsModule.forRoot([WorkspaceState])],
                    schemas: [NO_ERRORS_SCHEMA],
                    providers: [],
                });
            };

            configureTests(configure).then((testBed) => {
                fixture = testBed.createComponent(CompositionPanelComponent);
                store = testBed.get(Store);
            });
        })
    );

    it('When PolicyInstance Selected => Expect (info, policyTargets and policyProperties) tabs appear', () => {

        const testInstance = new PolicyInstance();

        fixture.componentInstance.initTabs(testInstance);

        expect (fixture.componentInstance.tabs.length).toBe(3);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.policyTargets);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.policyProperties);
    });

    it('should match current snapshot of composition-panel component.', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('When Topology Template is Service and no instance is selected Expect (info, deployment, inputs, info and api)', () => {

        const selectedComponent: Service = new Service(null, null);
        selectedComponent.isResource = jest.fn(() => false);
        selectedComponent.isService = jest.fn(() => true );

        fixture.componentInstance.store.select = jest.fn(() => Observable.of(selectedComponent));

        // const pnfMock = Mock.of<Service>({ isResource : () => false });
        fixture.componentInstance.topologyTemplate = selectedComponent;

        // Call ngOnInit
        fixture.componentInstance.ngOnInit();

        // Expect that
        expect (fixture.componentInstance.tabs.length).toBe(5);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.deploymentArtifacts);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.inputs);
        expect (fixture.componentInstance.tabs[3]).toEqual(tabs.infoArtifacts);
        expect (fixture.componentInstance.tabs[4]).toEqual(tabs.apiArtifacts);

    });

    it('When Topology Template is Resource and no instance is selected Expect (info, deployment, inputs, info and api)', () => {

        const selectedComponent: Service = new Service(null, null);
        selectedComponent.isResource = jest.fn(() => true);
        selectedComponent.isService = jest.fn(() => false );

        fixture.componentInstance.store.select = jest.fn(() => Observable.of(selectedComponent));

        fixture.componentInstance.topologyTemplate = selectedComponent;

        // Call ngOnInit
        fixture.componentInstance.ngOnInit();

        // Expect that
        expect (fixture.componentInstance.tabs.length).toBe(5);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.deploymentArtifacts);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.properties);
        expect (fixture.componentInstance.tabs[3]).toEqual(tabs.infoArtifacts);
        expect (fixture.componentInstance.tabs[4]).toEqual(tabs.reqAndCapabilities);

    });

    it('When Topology Template is Service and proxyService instance is selected ' +
        'Expect (info, deployment, inputs, info and api)', () => {

        const selectedComponent: Service = new Service(null, null);
        selectedComponent.isResource = jest.fn(() => false);
        selectedComponent.isService = jest.fn(() => true );

        fixture.componentInstance.store.select = jest.fn(() => Observable.of(selectedComponent));
        fixture.componentInstance.selectedComponentIsServiceProxyInstance = jest.fn(() => true);

        // const pnfMock = Mock.of<Service>({ isResource : () => false });
        fixture.componentInstance.topologyTemplate = selectedComponent;

        // Call ngOnInit
        fixture.componentInstance.ngOnInit();

        // Expect that
        expect (fixture.componentInstance.tabs.length).toBe(6);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.properties);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.reqAndCapabilities);

    });

    it('When Topology Template is Resource and VL is selected ' +
        'Expect (info, deployment, inputs, info and api)', () => {

        const topologyTemplate: Resource = new Resource(null, null);
        topologyTemplate.isResource = jest.fn(() => true);
        topologyTemplate.isService = jest.fn(() => false );

        const vlMock = Mock.of<Resource>({ resourceType : 'VL', isResource : () => true, isService : () => false });
        fixture.componentInstance.store.select = jest.fn(() => Observable.of(vlMock));

        fixture.componentInstance.topologyTemplate = topologyTemplate;

        // Call ngOnInit
        fixture.componentInstance.ngOnInit();

        // Expect that
        expect (fixture.componentInstance.tabs.length).toBe(5);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.deploymentArtifacts);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.properties);
        expect (fixture.componentInstance.tabs[3]).toEqual(tabs.infoArtifacts);
        expect (fixture.componentInstance.tabs[4]).toEqual(tabs.reqAndCapabilities);

    });

    it('When Topology Template is Service and VL is selected ' +
        'Expect (info, deployment, inputs, info and api)', () => {

        const topologyTemplate: Service = new Service(null, null);
        topologyTemplate.isResource = jest.fn(() => true);
        topologyTemplate.isService = jest.fn(() => false );

        const vlMock = Mock.of<Resource>({ resourceType : 'VL', isResource : () => true, isService : () => false });
        fixture.componentInstance.store.select = jest.fn(() => Observable.of(vlMock));

        fixture.componentInstance.topologyTemplate = topologyTemplate;

        // Call ngOnInit
        fixture.componentInstance.ngOnInit();

        // Expect that
        expect (fixture.componentInstance.tabs.length).toBe(5);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.deploymentArtifacts);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.properties);
        expect (fixture.componentInstance.tabs[3]).toEqual(tabs.infoArtifacts);
        expect (fixture.componentInstance.tabs[4]).toEqual(tabs.reqAndCapabilities);

    });

    it('When GroupInstance Selected => Expect (info, groupMembers and groupProperties) tabs appear.', () => {

        const testInstance = new GroupInstance();
        fixture.componentInstance.initTabs(testInstance);

        expect (fixture.componentInstance.tabs.length).toBe(3);
        expect (fixture.componentInstance.tabs[0]).toEqual(tabs.infoTab);
        expect (fixture.componentInstance.tabs[1]).toEqual(tabs.groupMembers);
        expect (fixture.componentInstance.tabs[2]).toEqual(tabs.groupProperties);
    });

});
