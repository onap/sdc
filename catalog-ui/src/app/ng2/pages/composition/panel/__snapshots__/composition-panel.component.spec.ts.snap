// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition-panel component should match current snapshot of composition-panel component. 1`] = `
<ng2-composition-panel
  activatePreviousActiveTab={[Function Function]}
  classes={[Function String]}
  initTabs={[Function Function]}
  isComponentInstanceSelected={[Function Function]}
  isConfiguration={[Function Function]}
  isPNF={[Function Function]}
  selectedComponentIsServiceProxyInstance={[Function Function]}
  selectedComponentIsServiceSubstitutionInstance={[Function Function]}
  selectedComponentIsVfcInstance={[Function Function]}
  setActive={[Function Function]}
  store={[Function Store]}
  toggleSidebarDisplay={[Function Function]}
>
  
</ng2-composition-panel>
`;
